reference-library/BeforeSwapDelta.md

```diff
@@ -0,0 +1,218 @@
+
+## Type Definition
+
+```solidity
+type BeforeSwapDelta is int256;
+```
+
+The `BeforeSwapDelta` type is an alias for int256, where:
+
+- The upper 128 bits represent the delta in specified tokens
+- The lower 128 bits represent the delta in unspecified tokens
+
+## Using Directives
+
+```solidity
+using BeforeSwapDeltaLibrary for BeforeSwapDelta global;
+using SafeCast for int256;
+```
+
+These using directives enable library functions to be used directly on `BeforeSwapDelta` values and provide safe casting operations for int256 values.
+
+## Functions
+
+### toBeforeSwapDelta
+
+```solidity
+function toBeforeSwapDelta(int128 deltaSpecified, int128 deltaUnspecified) pure returns (BeforeSwapDelta beforeSwapDelta);
+```
+
+Creates a `BeforeSwapDelta` value from two `int128` values representing `deltaSpecified` and `deltaUnspecified`.
+
+| Param Name | Type    | Description                          |
+|------------|---------|--------------------------------------|
+| deltaSpecified    | int128  | The balance change for the specified token       |
+| deltaUnspecified    | int128  | The balance change for the unspecified token      |
+
+Returns the created `BeforeSwapDelta` value.
+
+This function uses bitwise operations in assembly for gas-efficient packing of the two int128 values:
+
+```solidity
+assembly ("memory-safe") {
+    beforeSwapDelta := or(shl(128, deltaSpecified), and(sub(shl(128, 1), 1), deltaUnspecified))
+}
+```
+
+## Library Functions
+
+### ZERO_DELTA
+
+```solidity
+BeforeSwapDelta public constant ZERO_DELTA = BeforeSwapDelta.wrap(0);
+```
+
+A constant representing a zero delta (no balance changes).
+
+### getSpecifiedDelta
+
+```solidity
+function getSpecifiedDelta(BeforeSwapDelta delta) internal pure returns (int128 deltaSpecified);
+```
+
+Extracts the specified token delta from a `BeforeSwapDelta` value.
+
+| Param Name | Type         | Description                          |
+|------------|--------------|--------------------------------------|
+| delta      | BeforeSwapDelta | The `BeforeSwapDelta` value       |
+
+Returns the extracted specified token delta as an `int128`.
+
+### getUnspecifiedDelta
+
+```solidity
+function getUnspecifiedDelta(BeforeSwapDelta delta) internal pure returns (int128 deltaUnspecified);
+```
+
+Extracts the unspecified token delta from a BeforeSwapDelta value.
+
+| Param Name | Type         | Description                          |
+|------------|--------------|--------------------------------------|
+| delta          | BeforeSwapDelta | The `BeforeSwapDelta`       |
+
+Returns the extracted unspecified token delta as an `int128`.
+
+## Usage in Hooks
+
+When a hook is called during a swap operation, it can perform custom logic and interact with the pool's balances. The `beforeSwap` hook returns a `BeforeSwapDelta` value to indicate any balance changes it introduces. This value can then be forwarded to the `afterSwap` hook, allowing for proper accounting and balance reconciliation.
```

Comment:
```suggestion
When a hook is called during a swap operation, it can perform custom logic and interact with the pool's balances. The `beforeSwap` hook returns a `BeforeSwapDelta` value to indicate any balance changes the *hook* introduces. For example, hooks taking fees should return the value it took as a `BeforeSwapDelta`.
```